components/brave_federated_learning/brave_operational_profiling.cc

```diff
@@ -0,0 +1,201 @@
+    : pref_service_(pref_service), url_loader_factory_(url_loader_factory) {}
+
+BraveOperationalProfiling::~BraveOperationalProfiling() {}
+
+void BraveOperationalProfiling::RegisterLocalStatePrefs(
+    PrefRegistrySimple* registry) {
+  registry->RegisterIntegerPref(kLastCheckedSlotPrefName, -1);
+  registry->RegisterStringPref(kEphemeralIDPrefName, {});
+  registry->RegisterTimePref(kEphemeralIDExpirationPrefName, base::Time());
+}
+
+void BraveOperationalProfiling::Start() {
+  DCHECK(!simulate_local_training_step_timer_);
+  DCHECK(!collection_slot_periodic_timer_);
+
+  LoadPrefs();
+
+  simulate_local_training_step_timer_ =
+      std::make_unique<base::RetainingOneShotTimer>();
+  simulate_local_training_step_timer_->Start(
+      FROM_HERE, base::TimeDelta::FromSeconds(fake_update_duration_ * 60), this,
+      &BraveOperationalProfiling::OnSimulateLocalTrainingStepTimerFired);
+
+  collection_slot_periodic_timer_ = std::make_unique<base::RepeatingTimer>();
+  collection_slot_periodic_timer_->Start(
+      FROM_HERE, base::TimeDelta::FromSeconds(collection_slot_size_ * 60 / 2),
+      this, &BraveOperationalProfiling::OnCollectionSlotStartTimerFired);
+}
+
+void BraveOperationalProfiling::LoadParams() {
+  collection_slot_size_ =
+      operational_profiling::features::GetCollectionSlotSizeValue();
+  fake_update_duration_ = operational_profiling::features::
+      GetSimulateLocalTrainingStepDurationValue();
+  ephemeral_id_lifetime_ =
+      operational_profiling::features::GetEphemeralIdLifetime();
+
+  operational_profiling_endpoint_ = GURL(federatedLearningUrl);
+
+  LoadPrefs();
+  reuseOrRefreshEphemeralId();
+}
+
+void BraveOperationalProfiling::LoadPrefs() {
+  platform_ = GetPlatformIdentifier();
+  last_checked_slot_ = pref_service_->GetInteger(kLastCheckedSlotPrefName);
+  ephemeral_id_ = pref_service_->GetString(kEphemeralIDPrefName);
+  ephemeral_id_expiration_ =
+      pref_service_->GetTime(kEphemeralIDExpirationPrefName);
+}
+
+void BraveOperationalProfiling::SavePrefs() {
+  pref_service_->SetInteger(kLastCheckedSlotPrefName, last_checked_slot_);
+  pref_service_->SetString(kEphemeralIDPrefName, ephemeral_id_);
+  pref_service_->SetTime(kEphemeralIDExpirationPrefName,
+                         ephemeral_id_expiration_);
+}
+
+void BraveOperationalProfiling::OnCollectionSlotStartTimerFired() {
+  simulate_local_training_step_timer_->Reset();
+}
+
+void BraveOperationalProfiling::OnSimulateLocalTrainingStepTimerFired() {
+  SendCollectionSlot();
+}
+
+void BraveOperationalProfiling::SendCollectionSlot() {
+  current_collected_slot_ = GetCurrentCollectionSlot();
+  if (current_collected_slot_ == last_checked_slot_) {
+    return;
+  }
+
+  reuseOrRefreshEphemeralId();
+
+  auto resource_request = std::make_unique<network::ResourceRequest>();
+  resource_request->url = operational_profiling_endpoint_;
+  resource_request->headers.SetHeader("X-Brave-Trace", "?1");
```

Comment:
should we change header name to something like `X-Brave-FL-Operational-Profile`

Reply:
good point 👍